core/embed/trezorhal/entropy.c

```diff
@@ -0,0 +1,74 @@
+/*
+ * This file is part of the Trezor project, https://trezor.io/
+ *
+ * Copyright (c) SatoshiLabs
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "entropy.h"
+
+#include <string.h>
+
+#include "common.h"
+#include "flash.h"
+#include "rand.h"
+
+#include "stm32f4xx_ll_utils.h"
+
+extern __IO uint32_t uwTick;
+
+uint8_t HW_ENTROPY_DATA[HW_ENTROPY_LEN];
+uint8_t SW_ENTROPY_DATA[SW_ENTROPY_LEN] = {0};
+static size_t sw_entropy_index = {0};
+
+void collect_hw_entropy(void) {
+  // collect entropy from UUID
+  uint32_t w = LL_GetUID_Word0();
+  memcpy(HW_ENTROPY_DATA, &w, 4);
+  w = LL_GetUID_Word1();
+  memcpy(HW_ENTROPY_DATA + 4, &w, 4);
+  w = LL_GetUID_Word2();
+  memcpy(HW_ENTROPY_DATA + 8, &w, 4);
+
+  // set entropy in the OTP randomness block
+  if (secfalse == flash_otp_is_locked(FLASH_OTP_BLOCK_RANDOMNESS)) {
+    uint8_t entropy[FLASH_OTP_BLOCK_SIZE];
+    random_buffer(entropy, FLASH_OTP_BLOCK_SIZE);
+    ensure(flash_otp_write(FLASH_OTP_BLOCK_RANDOMNESS, 0, entropy,
+                           FLASH_OTP_BLOCK_SIZE),
+           NULL);
+    ensure(flash_otp_lock(FLASH_OTP_BLOCK_RANDOMNESS), NULL);
+  }
+  // collect entropy from OTP randomness block
+  ensure(flash_otp_read(FLASH_OTP_BLOCK_RANDOMNESS, 0, HW_ENTROPY_DATA + 12,
+                        FLASH_OTP_BLOCK_SIZE),
+         NULL);
+}
+
+void add_sw_entropy(uint8_t *data, size_t data_length) {
```

Comment:
After seeing that we automatically reseed from the entropy pool every 1 second, maybe entropy diffusion is not necessary. Do we have some statistics about the maximum amount of data that gets added to the entropy pool in 1 second?

Reply:
To be honest I didn't think much about feeding the drbg with software entropy. I considered it as a source of entropy in case the trng fails or have low entropy.

It cannot protect against an adversary that have physical access to the device for two reasons:
  * The adversary (as you've already said) is able to generate all the events themself.
  * Since the entropy of the pool that is mixed into the drbg every second is very low, the adversary is able to bruteforce the content of the pool.

Yarrow has two entropy pools (they call them slow and fast), each pool has an entropy threshold (that is higher for the slow pool). Each pool estimates the entropy it holds and when the entropy hits the threshold, the pool is mixed into the drbg.

Fortuna has 32 entropy pools. The drbg is reseeded when the first pool is full (but at at least 100 ms after the last reseed). The pool `i` is used once every `2^i` reseedings. (Multiple pools can be used in one reseeding.)

The disadvantage of Yarrow is that it needs the entropy estimator (that is a kind of black magic I suppose). The disadvantage of Fortuna is that is needs a lot of space and is rather inefficient (recovery time of Fortuna is something like 64 times longer than the recovery time of Yarrow with an ideal estimator).

Reply:
Let's not worry about multiple entropy pools then. But do we have some statistics about the maximum amount of data that gets added to the entropy pool in 1 second?

Reply:
No but it's not difficult to collect them. But what do you need them for? What is "maximum" for you? Is that "maximum" during "normal usage"? That depends also on how fast you can type on the touchscreen.

Reply:
> But what do you need them for?

The XOR operation provides no diffusion. One should avoid XOR-ing entropy twice to the same location in the pool before some form of entropy diffusion occurs, because otherwise you are likely to lose a significant part of the gathered entropy. In this implementation the entropy diffusion is effectively done by the periodic reseeds which occur every 1000 ms. The solution is to either have a large enough entropy pool so that no location gets written to twice between reseeds or to diffuse the entropy every time it's added to the pool. As I mentioned in the initial comment, the diffusion doesn't need to be cryptographically safe nor does it need to diffuse over the entire pool. Making the pool consist of, say, 4 to 8 uint32_t slots and using Jenkins hash like Fortuna would seem perfectly sufficient for this purpose.

> What is "maximum" for you? Is that "maximum" during "normal usage"?

Maximum during peak usage.